### PR TITLE
filter list of entries

### LINKED ISSUE #46: Search / filter list of saved entries
## Acceptance criteria
- Typing in the search bar dynamically changes the list of entries
- Search should filter the URL, username, and entry/title
- The results should appear in the order that they're returned from the database (no additional support for organizing results at this time)
- the cancel button appears in the search bar when text has been entered
- tapping the cancel button clears the current search text & restores the full list of entries

## Visual Design
**Invision**
Default: https://mozilla.invisionapp.com/share/XCEV238ZD#/267560598_Entry_List
Filtered: https://mozilla.invisionapp.com/share/XCEV238ZD#/283633648_Entry_List_-_Searching

**Zeplin**
Default: https://zpl.io/aBOOd4L
Filtered: https://zpl.io/aRNpKpN

### PR DESCRIPTION
**Pull request by sashei**
resolves #46

### CODE REVIEW

File: lockbox-ios/Presenter/ItemListPresenter.swift [hunk 1]
```diff
@@ -38,17 +54,33 @@ class ItemListPresenter {
...
+                    return !latest.items.isEmpty
                 }
+                .map { (latest: ItemsText) -> [Item] in
+                    if latest.text.isEmpty {
+                        return latest.items
+                    }
+
+                    let latestText = latest.text.lowercased()
+                    return latest.items.filter { item -> Bool in
+                        return item.entry.username?.lowercased().contains(latestText) ?? false
```
Comment (jhugman) [inline lockbox-ios/Presenter/ItemListPresenter.swift]: Be really wary about this — if your text is user-enterable, then you can't assume converting to lowercase and comparing will be correct.

Instead, use:
```swift
a.caseInsensitiveCompare(b) == .orderedSame
```

Comment (jhugman) [reply lockbox-ios/Presenter/ItemListPresenter.swift]: I'd also group the optional unwrapping together: 

```swift
guard let itemText = item.entry.username ?? item.origins.first ?? item.title else {
    return false
}
return itemText.caseInsensitiveCompare(latest.text) == .orderedSame
```

File: lockbox-ios/Presenter/ItemListPresenter.swift [hunk 2]
```diff
@@ -13,11 +13,23 @@ protocol ItemListViewProtocol: class {
     func hideEmptyStateMessaging()
 }
 
+struct ItemsText {
```
Comment (jhugman) [inline lockbox-ios/Presenter/ItemListPresenter.swift]: Nit: naming. 

Having that `s` in the middle will get old fast.

Suggest, perhaps: `NamedItemList`, or `TextAndItemList`.

File: lockbox-ios/Storyboard/Base.lproj/ItemList.storyboard [hunk 3]
```diff
@@ -77,7 +114,10 @@
                 </viewController>
                 <placeholder placeholderIdentifier="IBFirstResponder" id="WLg-fp-HvJ" userLabel="First Responder" sceneMemberID="firstResponder"/>
             </objects>
-            <point key="canvasLocation" x="750" y="197"/>
+            <point key="canvasLocation" x="749.60000000000002" y="196.55172413793105"/>
```
Comment (jhugman) [inline lockbox-ios/Storyboard/Base.lproj/ItemList.storyboard]: Really?

File: lockbox-ios/View/ItemListView.swift [hunk 4]
```diff
@@ -76,17 +85,44 @@ extension ItemListView: ItemListViewProtocol {
...
-                    guard let cell = tableView.dequeueReusableCell(withIdentifier: "itemlistcell") as? ItemListCell else { // swiftlint:disable:this line_length
-                        fatalError("couldn't find the right cell!")
+        self.dataSource = RxTableViewSectionedAnimatedDataSource<ItemSectionModel>(
+                configureCell: { dataSource, tableView, path, _ in
+                    let cellConfiguration = dataSource[path]
+
+                    var retCell: UITableViewCell
+                    switch cellConfiguration {
+                    case .Search:
+                        guard let cell = tableView.dequeueReusableCell(withIdentifier: "filtercell") as? FilterCell,
```
Comment (jhugman) [inline lockbox-ios/View/ItemListView.swift]: Are these cells `register`ed with the tableView anywhere?

File: lockbox-ios/View/ItemListView.swift [hunk 5]
```diff
@@ -76,17 +85,44 @@ extension ItemListView: ItemListViewProtocol {
...
+                        cell.filterTextField.rx.text
+                                .orEmpty
+                                .asObservable()
+                                .bind(to: presenter.filterTextObserver)
+                                .disposed(by: cell.disposeBag)
+
+                        retCell = cell
+                    case .Item(let title, let username, _):
+                        guard let cell = tableView.dequeueReusableCell(withIdentifier: "itemlistcell") as? ItemListCell else { // swiftlint:disable:this line_length
+                            fatalError("couldn't find the right cell!")
```
Comment (jhugman) [inline lockbox-ios/View/ItemListView.swift]: Ug. Fatal Errors make me queasy.

File: lockbox-ios/View/ItemListView.swift (hunk 4 above)
Comment (sashei) [reply lockbox-ios/View/ItemListView.swift]: storyboard.

File: lockbox-ios/Presenter/ItemListPresenter.swift (hunk 1 above)
Comment (sashei) [reply lockbox-ios/Presenter/ItemListPresenter.swift]: `.caseInsensitiveCompare()` does a complete comparison; we should return matches even if the entered text only matches a substring of the username or title or origin.

Comment (jhugman) [reply lockbox-ios/Presenter/ItemListPresenter.swift]: Gah! I misread. 

Nevertheless: we shouldn't be rolling our own case insensitive string ops. They aren't locale aware; there are alphabets that don't have a 1-1 mapping between upper and lower cases. 

Strings are weird. Rely on the OS (via [ICU][1]) to do the work for you.

```swift
guard let itemText = item.entry.username ?? item.origins.first ?? item.title else {
    return false
}
return itemText.localizedCaseInsensitiveContains(latest.text)
```

[1]: http://site.icu-project.org/

**Pull request merged**